catalog/core/catalog-core-api-impl/src/main/java/ddf/catalog/util/impl/QueryResultPaginator.java

```diff
@@ -0,0 +1,186 @@
+     * and startIndex.
+     * <p>
+     * The resultsRemaining are the total results not yet returned in a collection of the size
+     * of a page. This value is tracked through the hasNextPage() method to inform the client that
+     * there are valid results to continue querying.
+     *
+     * @param catalogFramework
+     * @param query
+     */
+    public QueryResultPaginator(CatalogFramework catalogFramework, QueryImpl query) {
+
+        Validate.notNull(catalogFramework);
+        Validate.notNull(query);
+
+        this.catalogFramework = catalogFramework;
+        this.queryCopy = copy(query);
+
+        isExhausted = false;
+        pageSize = queryCopy.getPageSize();
+        indexOffset = queryCopy.getStartIndex();
+
+        Validate.inclusiveBetween(1,
+                Integer.MAX_VALUE,
+                indexOffset,
+                "Query request start index must be greater than zero");
+
+        queriedResultsBuffer = new LinkedBlockingDeque<>();
+    }
+
+    // TODO: 6/27/17 Move this method over to QueryResultIterable inner class
+    /**
+     * If there are pages beyond the current page, returns true.
+     *
+     * @return True if there are results beyond the current page
+     */
+    /**
+     * @return true if there are more results for the client
+     */
+    public boolean hasNext() {
+
+        if (notEmpty()) {
+             return true;
+         }
+
+         if (isExhausted) {
+             return false;
+         }
+
+         //Buffer is empty, but query might not be exhausted.
+         fetchNext();
+         return notEmpty();
+    }
+
+    protected boolean notEmpty() {
+        return !queriedResultsBuffer.isEmpty();
+    }
+
+    /**
+     * Catalog Framework queried for results until current page is full of results
+     * or endIndex is reached.
+     *
+     * @return Collection of results
+     * @throws NoSuchElementException if the iteration has no more elements
+     */
+    public List<Result> next() throws NoSuchElementException {
+        if (!hasNext()) {
+            throw new NoSuchElementException();
+        }
+
+        while (!isExhausted && getBufferSize() < pageSize) {
+            fetchNext();
+        }
+
+        List<Result> queriedResultsBufferCopy = new ArrayList<>(pageSize);
+        queriedResultsBuffer.drainTo(queriedResultsBufferCopy, pageSize);
+        return queriedResultsBufferCopy;
+    }
+
+    /**
+     * Get the next batch of results.
+     * Assume the current index is pointed to the first result
+     * that has not been retrieved.
+     * Update the current index after the results are fetched.
+     * Add the fetched results to the queriedResultsBuffer.
+     * If no results were returned, assume there are no more results possible.
+     */
+    private void fetchNext() {
+
+        queryCopy.setStartIndex(indexOffset);
+        List<Result> resultList = queryCatalogFramework();
```

Comment:
❓ Is there any reason not to just use the list of results returned by the framework as the backing data structure for each chunk instead of dumping them into an intermediate BlockingQueue and then pulling them back out and copying them back into another list?